### PR TITLE
sysext: Fix issues with merged hierarchy mode

### LINKED ISSUE #32217: some sysext review points need addressing
### systemd version the issue has been seen with

git main

### Used distribution

_No response_

### Linux kernel version used

_No response_

### CPU architectures issue was seen on

None

### Component

_No response_

### Expected behaviour you didn't see

Just a tracker for the post-merge review points listed here:

https://github.com/systemd/systemd/pull/31801#pullrequestreview-1993682430

### Unexpected behaviour you saw

_No response_

### Steps to reproduce the problem

_No response_

### Additional program output to the terminal or log subsystem illustrating the issue

_No response_

### PR DESCRIPTION
**Pull request by krnowak**
There seems to be a regression, where merged hierarchy ends up having `0700` permissions, which breaks boot on Flatcar. No binary under `/usr` can be executed - we are getting `permission denied` errors.

The PR:
- fixes some small error reporting issues,
- makes sysext to ensure that the merged hierarchy has the same mode as the underlying hierarchy
- makes sysext to require that the preexisting mutable directory has the same mode as the underlying hierarchy
- adds tests
- makes the tests to run also on `/` (so without the `--root` flag)
- checks for used extensions before potentially creating the mutable directory when in `--mutable=yes` mode
  - so we will skip creating the directory if we there are no extensions for the hierarchy at all thus no overlay will be created
- addresses post-merge review issues from https://github.com/systemd/systemd/pull/31801#pullrequestreview-1993682430
- closes #32217

### CODE REVIEW

**Comment by krnowak** [comment]
Currently building a Flatcar image with those patches to see if they fix the booting problem. If they do, I'll mark this PR as ready for review.

**Comment by krnowak** [comment]
The failures in CI seem to be unrelated and of the same nature - not able to install same debian test dependencies.

Also, the PR seems to have a missing `sysext` label.

**Comment by pothos** [comment]
(FYI should be merged before the next release)

**Comment by pothos** [comment]
The issue this PR fixes was originally reported in https://github.com/systemd/systemd/pull/31000#issuecomment-1948455247 and I forgot to check it again before the PR got merged.
I've tested the read-only (default) merge and it now works, but for `--mutable=yes` I get this new error:
```
Assertion 'root' failed at src/sysext/sysext.c:780, function mutable_directory_mode_matches_hierarchy(). Aborting.
```

**Comment by pothos** [comment]
One more small issue I ran into is that the `ephemeral-import` only works after `mkdir -p /var/lib/extensions.mutable/opt` but not when the folder doesn't exist - I think because the test sysext only ships `/usr` and thus there is no lowerdir for `/opt`. But ideally this would still work because in the end no overlay mount for `/opt` is set up even after creating `/var/lib/extensions.mutable/opt`.
```
# systemd-sysext refresh --mutable=ephemeral-import
Using extensions 'test'.
Unmerged '/usr'.
Not importing mutable directory for hierarchy /opt as a lower dir, because it points to the hierarchy itself
Failed to merge hierarchies
```

**Comment by yuwata** [comment]
Sorry, please rebase.

File: src/sysext/sysext.c [hunk 1]
```diff
@@ -768,8 +768,38 @@ static int resolve_hierarchy(const char *hierarchy, char **ret_resolved_hierarch
...
+        if (stat(path_in_root, &st) < 0) {
+                if (errno == ENOENT)
+                        return 0;
+                return log_error_errno(errno, "Failed to stat mutable directory '%s': %m", path_in_root);
+        }
+
+        actual_mode = st.st_mode & 0777;
+        if (actual_mode != hierarchy_mode) {
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Mutable directory '%s' has mode %04o, ought to have mode %04o", path_in_root, actual_mode, hierarchy_mode);
+        }
```
Comment (poettering) [inline src/sysext/sysext.c]: no {} around single line if blocks, as per coding style.

File: src/sysext/sysext.c [hunk 2]
```diff
@@ -962,8 +962,13 @@ static int maybe_import_ignored_mutable_directory(OverlayFSPaths *op) {
...
 
         r = chase(path, arg_root, CHASE_PREFIX_ROOT, &resolved_path, NULL);
-        if (r < 0 && r != -ENOENT)
+        if (r < 0) {
+                if (r == -ENOENT) {
+                        log_debug("Mutable directory for %s does not exist, not importing", op->hierarchy);
+                        return 0;
+                }
                 return log_error_errno(r, "Failed to resolve mutable directory '%s': %m", path);
+        }
```
Comment (poettering) [inline src/sysext/sysext.c:986]: i would format that as:

```c
if (r == -ENOENT) {
        …
}
if (r < 0)
        …
```

but doesn't matter.

(shallow nesting is best nesting)

**Comment by poettering** [comment]
lgtm. some minor nitpicks

**Comment by pothos** [comment]
Thanks for the fixes, I've retested it and now it works with `--mutable=yes` and `ephemeral-import` even without creating the `opt` subdir.

File: src/basic/path-util.c [hunk 3]
```diff
@@ -525,6 +525,18 @@ int path_compare_filename(const char *a, const char *b) {
         return strcmp(fa, fb);
 }
 
+int path_equal_or_inode_same_checked(const char *a, const char *b, int flags) {
```
Comment (YHNdnzj) [inline src/basic/path-util.c]: Please call this `path_equal_or_inode_same_full` or so. "checked" is harder to understand.

**Comment by pothos** [comment]
> I'm not sure if anything should be done with it, but maybe we could say that if the extension has only `/usr/lib/extension-release.d/extension-release.${name}` then `/usr/` will be overlaid, but as readonly in such case. In order to have it mutable too, create some other dummy file under `/usr/`. This maybe would need some more changes in uapi extensions draft to allow implementations to decide to create a read-only overlay in certain situations.

Having different behavior based on the contents of extensions is something I would want to avoid because it leads to surprising behavior. Ideally we would even have a way of always setting up the mountpoint to further align the behavior. Whether `/usr`and `/opt` become mutable together or separate is something I would expect to be a new flag (and yes, I'm still convinced about the need for having a config file because that's the only sane way we can set up the overlay from the initrd the same way the final system would do, because otherwise one would have to parse the drop-in units for the env vars…).

File: src/sysext/sysext.c (hunk 1 above)
Comment (krnowak) [reply src/sysext/sysext.c]: Done.

File: src/sysext/sysext.c (hunk 2 above)
Comment (krnowak) [reply src/sysext/sysext.c:986]: Fixed anyway.

File: src/basic/path-util.c (hunk 3 above)
Comment (krnowak) [reply src/basic/path-util.c]: Done.

**Comment by krnowak** [comment]
Rebased.

**Pull request merged**